pkg/daemon/httpchecker.go

```diff
@@ -25,27 +26,36 @@ import (
 )
 
 // WaitForServerReady waits until the HTTP(S) server can respond to a GET request. Should NOT allow cancelling the retry as shuting down non-ready server may cause deadlock.
-func WaitForServerReady(serverAddr string, insecureSkipVerify bool) error {
+func WaitForServerReady(serverAddr string, insecureSkipVerify bool, clientCertEnabled bool) error {
 
 	t := http.DefaultTransport.(*http.Transport).Clone()
 	t.TLSClientConfig = &tls.Config{}
 	client := &http.Client{Transport: t}
 
-	var url string
+	var targetUrl string
 	if insecureSkipVerify {
 		t.TLSClientConfig.InsecureSkipVerify = true
-		url = "https://" + serverAddr
+		targetUrl = "https://" + serverAddr
 	} else {
-		url = "http://" + serverAddr
+		targetUrl = "http://" + serverAddr
 	}
 
 	get := func() error {
-		resp, err := client.Get(url)
-		if err == nil {
+		resp, err := client.Get(targetUrl)
+		if err != nil {
+			// if client certificate disabled, return ALL errors.
+			// if client certificate enabled, return ALL errors but exclude client certificate verification error.
+			errCause := errors.Unwrap(err)
+			if !clientCertEnabled || (errCause != nil && errCause.Error() != "remote error: tls: certificate required") {
+				return err
+			}
+		}
+
+		if resp != nil {
```

Comment:
I think it is enough to just check the `err != nil` and no need to check `resp != nil`.
```suggestion
```
- https://pkg.go.dev/net/http#Get
- [underlining method: NewRequestWithContext](https://cs.opensource.google/go/go/+/master:src/net/http/request.go;drc=c5adb8216968be46bd11f7b7360a7c8bde1258d9;bpv=1;bpt=1;l=889?q=NewRequest&ss=go%2Fgo&gsn=NewRequestWithContext&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fgo.googlesource.com%2Fgo%3Flang%3Dgo%3Fpath%3Dnet%2Fhttp%23func%2520NewRequestWithContext)

Reply:
When verifying that client certificate enabled, http.Get fails, so resp becomes nil.

Reply:
@fum1h1to how about the following? I think we should have different debug logs for the 2 `return nil` cases.

```go
	get := func() error {
		resp, err := client.Get(targetUrl)
		if err != nil {
			// if client certificate disabled, return ALL errors.
			// if client certificate enabled, return ALL errors but exclude client certificate verification error.
			errCause := errors.Unwrap(err)
			if !clientCertEnabled || errCause == nil || errCause.Error() != "remote error: tls: certificate required" {
				return err
			}
			log.Debugf("Server started at %s (can response certificate required error)", targetUrl)
			return nil
		}

		resp.Body.Close()
		log.Debugf("Server started at %s", targetUrl)
		return nil
	}
```